### ISSUE TITLE
Learning rate and momentum as shared scalars

### ISSUE THREAD

**Issue by EderSantana** [fchollet/keras#536]
I'd like to ask if we could make the optimizer's learning rate (`lr`) and momentum (`momentum`) into shared scalars. This way we could change their values during training with `.set_value` using custom rules.

I could work on the PR as long as you guys don't have a reason for not to.

**Comment by fchollet** [comment]
I would have issues with changing the definition of the learning rate. But if the PR is just about switching `lr` and `momentum` to shared scalars (hence modifiable after compilation) instead of scalars, without touching the logic, that's fine by me.

**Comment by entron** [comment]
This could be a very useful feature.

**Comment by dongb5** [comment]
So right now, the only way to change learning rate during training is 1) Save weights of the model. 2) Compile a new model with same architecture and different learning rate. 3) Load weights and continue training??

**Comment by wuaalb** [comment]
I think you can also create a shared scalar and pass it to the optimizer's initializer.
I tried something like this for changing momentum from a callback.
https://github.com/wuaalb/keras_extensions/blob/master/keras_extensions/callbacks.py

**Comment by EderSantana** [comment]
The problem is, since you don't recompile the graph it won't use your new learning rate or momentum, even if you change it with a callback.

**Comment by wuaalb** [comment]
Something like this doesn't work, or isn't what you're looking for?
```python
model = ...
optimizer = SGD(lr=shared_scalar(0.01))
model.compile(optimizer, loss)
...
optimizer.lr.set_value(0.02) # e.g. from a callback during fit()
```

Anyways, making them shared scalars by default sounds good..

**Comment by EderSantana** [comment]
That is exactly what I'm doing! What we were discussing was, should we make `lr` a `shared_scalar` by default or not. It seems that we don't have anything to lose with that.

**Comment by fchollet** [comment]
> What we were discussing was, should we make lr a shared_scalar by default or not? It seems that we don't have anything to lose with that.

Indeed. You would pass a float to the constructor and internally this float would be used as the initial value of a shared scalar. The logic of the optimizer itself would be unchanged.
